File: quickstart_guides/trees/python/levelorder_traversal.py
"""
Title: Levelorder traversal

Problem:
    Given a binary tree, return the level-order traversal of its nodes' values.

Execution: python levelorder_traversal.py
"""
from typing import List
import unittest


class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


def helper_util(lst, rst, result, level):
    # To add layer to the stack
    if len(result) < level:
        result.append([])

    # Keep tracking the left sub tree
    if lst is not None:
        result[level - 1].extend([lst.val])
        helper_util(lst.left, lst.right, result, level + 1)

    # Keep tracking the right sub tree
    if rst is not None:
        result[level - 1].extend([rst.val])
        helper_util(rst.left, rst.right, result, level + 1)

    return result


def levelorder_traversal(root: TreeNode) -> List[List[int]]:

    if root is None:
        return []

    result = [[root.val]]
    lst = root.left
    rst = root.right

    result = helper_util(lst, rst, result, 2)

    result = [x for x in result if len(x) != 0]

    return result


class TestLevelorderTraversal(unittest.TestCase):
    """Unit tests for levelorder_traversal."""

    def test_1(self):
        tn = TreeNode(3)
        tn.right = TreeNode(20)
        tn.left = TreeNode(9)
        tn.right.left = TreeNode(15)
        tn.right.right = TreeNode(7)
        self.assertEqual(levelorder_traversal(tn), [[3], [9, 20], [15, 7]])


if __name__ == "__main__":
    unittest.main()

File: quickstart_guides/trees/java/LevelorderTraversal.java
/*
 *   Title: Level-order traversal of binary tree
 *
 * Problem:
 *   Given a binary tree, return the level-order traversal of its nodes' values.
 *
 *   Execution: javac LevelorderTraversal.java && java LevelorderTraversal
 */
import java.util.*;


// A Binary Tree node
class Node {
    int data;
    Node left, right;

    Node(int value) {
        data = value;
        left = right = null;
    }
}


class LevelorderTraversal {
    Node root;

    public static List<List<Integer>> levelOrderTraversal(Node root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if(root == null) return res;

        Queue<Node> q = new LinkedList<Node>();
        q.add(root);

        while(q.isEmpty() == false) {
            List<Integer> currli = new ArrayList<Integer>();
            int qsize = q.size();
            for(int i = 0; i < qsize; i++) {
                Node curr = q.poll();
                currli.add(curr.data);
                if(curr.left != null) q.add(curr.left);
                if(curr.right != null) q.add(curr.right);
            }
            res.add(currli);
        }

        return res;
    }

    public static void main(String args[])
    {
        LevelorderTraversal tree = new LevelorderTraversal();
        tree.root = new Node(1);
        tree.root.left = new Node(2);
        tree.root.right = new Node(3);
        tree.root.left.left = new Node(4);
        tree.root.left.right = new Node(5);
        tree.root.right.left = new Node(6);
        tree.root.right.right = new Node(7);

        System.out.println(levelOrderTraversal(tree.root));

        System.out.println("Passed all test cases");
    }
}
